### ISSUE TITLE
Contact form and recaptcha generator

### ISSUE THREAD

**Issue by pvawser01** [katalyst/koi#319]
- Start Date: 2017-04-03
- Owner: PV
- RFC PR: 

# Summary

*Add generator to KOI to build default contact form with recaptcha protection*

# Motivation

*We have had a number of issues with different clients where the public contact form was being hit by bots attempting DOS attacks.*
*As all sites these days have a contact form, it could be useful to have a generator to implement a default contact form, with routes, controllers, models and views with recaptcha added for bot attack prevention.

# Detailed design

*A generator and some templates would need to be built to facilitate the generation of a contact form implementation. Routes, models, views and controllers could all be bootsraped as the implementation is generally standard across all projects.*

# Drawbacks

*??*

# Alternatives

*??*

# Unresolved questions

*??*

**Comment by Obversity** [comment]
I like the idea of a generator for a contact form, but it might be worth thinking about whether or not we can just improve the mailer generator to the point where it does 90% of the work for a contact form. A `--with-recapcha` flag for the mailer generator would be amazing.

**Comment by pvawser01** [comment]
@Obversity True that could work. I was thinking that a specific generator could also help with the whole ui set up too. Bootstrap a basic template for contact then Dave could weave his magic over it. Might be overkill. Just though it is usually just boilerplate. Even the view to have a basic form in place.

**Comment by Obversity** [comment]
Yeah, my thinking was that lots of our sites take submissions other than just the contact form, and we almost always use the mailer generator for bootstrapping front-end forms, so having it optionally generate recaptcha stuff would be great. It'd need to:

- add recaptcha_tags to the form
- add the recaptcha verification logic to the controller
- add the gem if necessary
- add the initializer if necessary 
- give instructions on how to get recaptcha keys from google for a client

**Comment by dbaines** [comment]
Would we make the API key something they could enter in to their Site Settings area or would it be better suited in `application.yml`?

**Comment by Obversity** [comment]
We usually use the application.yml for recaptcha keys I think, but I don't see why it couldn't be done in site settings instead, now that you mention it.

**Comment by jsidoryn** [comment]
Since this is done so much this is a great candidate to streamline. 

Thinking these are the things that could be autogenerated
• the form (with re-capture)
• thank you page
• email (optionally to the submitter and / or specified email address)
• admin - reverse chronological order (if not already) and configure table columns
• set up a dashboard widget (related to dashboard RFC)

**Comment by Obversity** [comment]
I'm still wary of adding a whole new generator that's so similar to an existing generator.  It's definitely worth seeing whether we can extend the mailer generator to account for most of our needs for a contact form.

**Comment by dbaines** [comment]
Yeah I'm definitely in favour of extending the mailer generator, since it already does most of that. 
As Bill mentioned we just need a way toggle a widget and recaptcha as flags.
I'd be in favour of having recaptcha by default though, and having a `--no-recaptcha` flag instead. 

>  (optionally to the submitter)   

And an option for this too would be good. Sometimes we need emails just to go to the admin. Sometimes we need it to go the admin AND the user. It's only a few minutes to configure that but I suppose that's a few minutes that we save each time.

**Comment by pvawser01** [comment]
@Obversity  Having a specific generator is no a bad way to go. If you look at the default requirements for a contact form, keeping it seperate keeps the generator with a single task rather than complicating an existing one.
Given that pretty much all oour sites have contact forms, is why I originally thought adding it as part of the boot strap template might be more suitable.

**Comment by dbaines** [comment]
I think the discussed requirements of a contact form extend directly in to mailers in general to be honest. Being able to add captcha, dashboard modules and multiple mail sends would directly benefit all mailers, not just contact forms. 

We could create a generator specifically for a ContactForm model, but I'm assuming it would just be a mailer generator anyway with some default fields and flags. 

```ruby  
rails generate koi:contact_form  
# actually runs:
rails generate koi:mailer ContactForm first_name:string last_name:string subject:string query:text --with-widget
```

That's not a silly idea, I often make helper functions that just run functions with a set of preset arguments. 

I'm not entirely sold that all our contact forms are the same though. I know every one I've made has some level of uniqueness to it.

**Comment by pvawser01** [comment]
@dbaines Agreed. However, I think there is enough commonality to create a base setup

**Comment by jsidoryn** [comment]
didn't even know we had a mailer generator!

i agree that it's pretty similar and the first step might be to add the extra flags etc. it opens the discussion of what would be useful things to add to the generator. would be very cool if we get generate items for the dashboard as this would lower the bar to creating them. 

if we then want to take it one step further we can then make a specific one for a contact page as @pvawser01 is saying. we probably don't need to make a call on this until we've finished extending the generators.

**Comment by pvawser01** [comment]
This has been added to the master branch. All mailers will have recaptcha by default. If you want to opt out generate using the --skip_recaptcha option. The only thing it does not add is the google recaptcha key values in application.yml. The keys are there just empty values.
